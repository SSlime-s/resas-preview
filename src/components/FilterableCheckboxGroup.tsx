"use client";

import { useCallback, useMemo, useState, type Key } from "react";
import React from "react";

import { CheckBox } from "./CheckBox";
import { FilterInput } from "./FilterInput";

interface Option<T extends Key> {
	label: string;
	value: T;
}
interface Props<T extends Key> {
	options: Readonly<Option<T>[]>;
	selected: Readonly<T[]>;
	onSelectChange: (selected: T[]) => void;
	onFilter(
		value: string,
		options: Readonly<Option<T>[]>
	): Readonly<Option<T>[]>;
}
export function FilterableCheckboxGroup<T extends Key>({
	options,
	selected,
	onSelectChange,
	onFilter,
}: Props<T>) {
	const [filter, setFilter] = useState("");
	const filteredOptions = useMemo(
		() => onFilter(filter, options),
		[filter, options, onFilter]
	);

	const onFilterChange = useCallback(
		(e: React.ChangeEvent<HTMLInputElement>) => {
			setFilter(e.target.value);
		},
		[]
	);

	const onChange = useCallback(
		(value: T) => (e: React.ChangeEvent<HTMLInputElement>) => {
			if (e.target.checked) {
				onSelectChange([...selected, value]);
			} else {
				onSelectChange(selected.filter((v) => v !== value));
			}
		},
		[selected, onSelectChange]
	);

	return (
		<div>
			<FilterInput value={filter} onChange={onFilterChange} />

			{filteredOptions.map(({ label, value }) => (
				<CheckBoxWithOnChange
					key={value}
					label={label}
					value={value}
					checked={selected.includes(value)}
					onChange={onChange}
				/>
			))}
		</div>
	);
}

const CheckBoxWithOnChange = React.memo(
	CheckBoxWithOnChangeRaw
) as typeof CheckBoxWithOnChangeRaw;
function CheckBoxWithOnChangeRaw<T>({
	label,
	value,
	checked,
	onChange,
}: {
	label: string;
	value: T;
	checked: boolean;
	onChange: (
		value: NoInfer<T>
	) => (e: React.ChangeEvent<HTMLInputElement>) => void;
}) {
	const handleChange = useMemo(() => onChange(value), [value, onChange]);

	return <CheckBox label={label} checked={checked} onChange={handleChange} />;
}
