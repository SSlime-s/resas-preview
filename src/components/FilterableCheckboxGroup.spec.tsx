import { render } from "@testing-library/react";

import { FilterableCheckboxGroup } from "./FilterableCheckboxGroup";

function setup() {
	const options = [
		{ label: "Foo", value: "foo" },
		{ label: "Bar", value: "bar" },
		{ label: "Baz", value: "baz" },
	];

	const onSelectChange = vi.fn();
	const onFilter = vi.fn(() => options);
	const result = render(
		<FilterableCheckboxGroup
			options={options}
			selected={["foo"]}
			onSelectChange={onSelectChange}
			onFilter={onFilter}
		/>
	);

	return { ...result, mock: { onSelectChange, onFilter } };
}

describe("Components/FilterableCheckboxGroup", () => {
	it("スナップショット", () => {
		const { container } = setup();

		expect(container).toMatchSnapshot();
	});

	it("スナップショット: no item", () => {
		const { container } = render(
			<FilterableCheckboxGroup
				options={[]}
				selected={[]}
				onSelectChange={vi.fn()}
				onFilter={vi.fn(() => [])}
			/>
		);

		expect(container).toMatchSnapshot();
	});
});
