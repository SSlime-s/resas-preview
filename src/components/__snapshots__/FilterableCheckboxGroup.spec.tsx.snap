// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Components/FilterableCheckboxGroup > スナップショット 1`] = `
<div>
  <div
    class="🐻-1825998508"
  >
    <input
      type="search"
      value=""
    />
    <div
      class="🐻-2655331464"
    >
      <label>
        <input
          checked=""
          type="checkbox"
        />
        Foo
      </label>
      <label>
        <input
          type="checkbox"
        />
        Bar
      </label>
      <label>
        <input
          type="checkbox"
        />
        Baz
      </label>
    </div>
  </div>
</div>
`;
